crates/cargo-test-support/src/registry.rs

```diff
@@ -611,16 +638,72 @@ impl HttpServer {
                     self.dl(&req)
                 }
             }
+            // publish
+            ("put", ["api", "v1", "crates", "new"]) => {
+                if !authorized(true) {
+                    return self.unauthorized(req)
+                }
+                if let Some(body) = &req.body {
+                    // Get the metadata of the package
+                    let (len, remaining)  = body.split_at(4);
+                    let json_len = u32::from_le_bytes(len.try_into().unwrap());
+                    let (json, remaining) = remaining.split_at(json_len as usize);
+                    let new_crate = serde_json::from_slice::<crates_io::NewCrate>(json).unwrap();
+                    // Get the `.crate` file
+                    let (len, remaining) = remaining.split_at(4);
+                    let file_len = u32::from_le_bytes(len.try_into().unwrap());
+                    let (file, _remaining) = remaining.split_at(file_len as usize);
+
+                    // Write the `.crate`
+                    let dst = self.dl_path.join(&new_crate.name).join(&new_crate.vers).join("download");
+                    t!(fs::create_dir_all(dst.parent().unwrap()));
+                    let mut f = t!(File::create(&dst));
+                    t!(f.write_all(file));
```

Comment:
Just a minor simplification:

```suggestion
                    t!(fs::write(&dst, file));
```